File: Data Structures & Algorithms/max-water-container/submission-3.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        currentMax = 0
        
        left, right = 0, len(heights)-1

        while left < right:
            currentAmount = min(heights[left], heights[right]) * (right - left)
            currentMax = max(currentMax, currentAmount)

            if heights[left] <= heights[right]:
                left += 1
            else:
                right -= 1

        return currentMax

File: Data Structures & Algorithms/balanced-binary-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isBalanced(TreeNode* root) {
        bool balanced = true;

        dfs(root, balanced);
        return balanced;
    }

    int dfs(TreeNode* root, bool& balanced)
    {
        if (root == nullptr)
        {
            return 0;
        }

        int leftDepth = dfs(root->left, balanced);
        int rightDepth = dfs(root->right, balanced);

        balanced = balanced && abs(leftDepth - rightDepth) < 2;

        return max(leftDepth, rightDepth) + 1;
    }
};
